openslides_backend/action/actions/resource/upload.py

```diff
@@ -0,0 +1,71 @@
+import base64
+import mimetypes
+from typing import Any, Dict
+
+from ....models.models import Resource
+from ....shared.exceptions import ActionException
+from ....shared.filters import And, FilterOperator
+from ...generics.create import CreateAction
+from ...util.default_schema import DefaultSchema
+from ...util.register import register_action
+from ...util.typing import ActionPayload
+from .delete import ResourceDelete
+
+
+@register_action("resource.upload")
+class MediafileUploadAction(CreateAction):
+    """
+    Action to upload a resourcefile.
+    The token-field acts as unique key
+    """
+
+    model = Resource()
+    schema = DefaultSchema(model).get_create_schema(
+        required_properties=["token", "organisation_id"],
+        additional_required_fields={
+            "file": {"type": "string"},
+            "filename": {"type": "string"},
+        },
+    )
+
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        filename_ = instance.pop("filename")
+        file_ = instance.pop("file")
+        instance["mimetype"] = mimetypes.guess_type(filename_)[0]
+        if instance["mimetype"] is None:
+            raise ActionException(f"Cannot guess mimetype for {filename_}.")
+        decoded_file = base64.b64decode(file_)
+        instance["filesize"] = len(decoded_file)
+        id_ = instance["id"]
+        mimetype_ = instance["mimetype"]
+        msg = self.media.upload_resource(file_, id_, mimetype_)
+        if msg:
+            raise ActionException(msg)
+        return instance
+
+    def get_updated_instances(self, payload: ActionPayload) -> ActionPayload:
+        tokens = [instance["token"] for instance in payload]
+        if len(tokens) != len(set(tokens)):
+            raise ActionException(
+                "It is not permitted to use the same token twice in a request."
+            )
+
+        for instance in payload:
+            results = self.datastore.filter(
+                self.model.collection,
+                And(
+                    FilterOperator("token", "=", instance["token"]),
+                    FilterOperator("organisation_id", "=", instance["organisation_id"]),
+                ),
+            )
+            if len(results) == 0:
+                continue
+            elif len(results) == 1:
+                id = next(iter(results))
+                self.execute_other_action(ResourceDelete, [{"id": id}])
+            else:
+                raise ActionException(
```

Comment:
I would prefer logging this as an error and deleting all resources with the same token. This enables auto-fixing of invalid data.

Reply:
I don't think that it is a good idea, to auto-fix such an error. This error should/could/may never occur, but when, we should detect the reason. Therefore the error will be logged now.
Now there is a logger defined in Action, which now is available in all Action-classes.